### PR TITLE
Add `override` option

### PR DESCRIPTION
**Pull request by hrsh7th**
This PR adds the `sources[n].override` option that enables to override custom source behavior.

For example, you can change `is_available` condition like this.

```lua
cmp.setup {

  sources = {
    {
      name = 'nvim_lsp',
      override = {
        is_available = function()
          return context.in_treesitter_capture('comment')
        end
      }
    }
  }

}
```

This PR has following breaking changes.

- Remove `config.sources[n].keyword_pattern` option.
- Remove `config.sources[n].trigger_characters` option.

### CODE REVIEW

**Comment by max397574** [comment]
just wondering
what else does this pr add?
because I can't imagine that all those 5.6k loc are for the `override` option only

**Comment by hrsh7th** [comment]
https://github.com/hrsh7th/nvim-kit

This PR adds the nvim-kit.

**Comment by wookayin** [comment]
> because I can't imagine that all those 5.6k loc are for the override option only

+1, I think adding nvim-kit should be a separate commit/PR (actually I'm not sure this is good to be merged to nvim-cmp core). Why not as a separate library or dependencies? I see many duplicates with plenary.nvim. It also seems a few more refactoring would be needed. It'd be greatly appreciated if @hrsh7th can please explain more rationale and benefits of having that merged to core.

**Comment by hrsh7th** [comment]
The reason for embedding nvim-kit in core is that avoiding breaking changes and new dependencies for users.

**Comment by hrsh7th** [comment]
I already use nvim-kit in some of the my codebases.

https://github.com/hrsh7th/cmp-core-example
https://github.com/hrsh7th/nvim-linkedit
https://github.com/hrsh7th/nvim-gtd
https://github.com/hrsh7th/nvim-insx

As the neovim core provides higher level APIs, it becomes less necessary, but is still needed at this point.

**Comment by hrsh7th** [comment]
@wookayin I don't think it's bad to add the nvim-kit to the nvim-cmp. Why do you think it's bad?

**Comment by hrsh7th** [comment]
@wookayin ping

**Comment by hrsh7th** [comment]
@wookayin Hi, Shouldn't I install nvim-kit in nvim-cmp?

**Comment by f3fora** [comment]
Why not split the PR? One for `override` option and another for `nvim-kit`?

File: lua/cmp/init_spec.lua [hunk 1]
```diff
@@ -0,0 +1,158 @@
+local async = require('cmp.utils.async')
```
Comment (hrsh7th) [inline lua/cmp/init_spec.lua:1]: use nvim-kit in spec.

**Comment by hrsh7th** [comment]
yes. I can't take time at the moment, and my motivation has decreased, so I closed it.

**Pull request closed without merging**